include/deal.II/matrix_free/tensor_product_kernels.h

```diff
@@ -3272,49 +3272,59 @@ namespace internal
     else if (dim == 1)
       {
         // gradient
-        result[0] = values[1] - values[0];
+        result[0] = Number3(values[1]) - Number3(values[0]);
```

Comment:
Why do you choose to cast both values separately?
```suggestion
        result[0] = Number3(values[1] - values[0]);
```

Reply:
Good question. I chose this way because we use a cast of `Number3(values[0])` a line below, so the goal was to tell the compiler to only cast `values[0]` once. As background, the underlying type of `values[0]` is typically `double` or `Point<dim>`, while is most often `Number3 = VectorizedArray<double>` or `Tensor<1, dim, VectorizedArray<double>>`, respectively. Hence, the cast involves a broadcast, which might come from memory or from a value already in registers.
This choice is up to debate, as my compiler still chooses to load `values[1]` and `values[0]` as scalar variables, build a scalar difference, and then broadcast it (as well as the scalar `values[0]`). I think we can pick both, and if you prefer, we go with the scalar difference.

Reply:
Yes, we need two casts either way. The difference is that on Intel, broadcast from memory consumes less precious execution resources than scalar load + bcast from register. Look at https://www.agner.org/optimize/instruction_tables.pdf page 308 for `VBROADCASTSD z,m64` (load from memory + broadcast), which executes at 2 instructions per cycle (RCP 0.5) on execution ports 2 and 3, whereas `VBROADCASTSD v,x` (broadcast from register) executes at 1 instruction per cycle (RCP 1) on execution port 5, which is the port also executing FMAs. Now this is Intel specific (I don't know if Sapphire Rapids still has this behavior), and AMD does it right, see e.g. page 134 for Zen 4 (RCP 0.5 both from memory and registers and not as disturbing for the execution pipes of arithmetic work), so we can stick with the better instruction. (Or let the compiler decide on the right instruction, it seems it will exchange them to whatever is deemed more beneficial anyway.)

Reply:
I now switched to the single cast as suggested.